src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,621 @@
+                    receiver: _factory.Local(_scope.ContextVariable),
+                    logger,
+                    MakeStoreLoggerArguments(logger.Parameters[0], targetSymbol, targetType, assignment, refAssignmentSourceIndex, targetIndex))
+            }, VariableRead(targetSymbol));
+        }
+
+        private bool TryGetLocalOrParameterInfo(BoundNode node, [NotNullWhen(true)] out Symbol? symbol, [NotNullWhen(true)] out TypeSymbol? type, [NotNullWhen(true)] out BoundExpression? indexExpression)
+        {
+            if (node is BoundLocal { LocalSymbol: var localSymbol })
+            {
+                symbol = localSymbol;
+                type = localSymbol.Type;
+                indexExpression = _factory.LocalId(localSymbol);
+                return true;
+            }
+
+            if (node is BoundParameter { ParameterSymbol: var parameterSymbol })
+            {
+                Debug.Assert(!parameterSymbol.IsDiscard);
+
+                symbol = parameterSymbol;
+                type = parameterSymbol.Type;
+                indexExpression = _factory.ParameterId(parameterSymbol);
+                return true;
+            }
+
+            symbol = null;
+            indexExpression = null;
+            type = null;
+            return false;
+        }
+
+        private ImmutableArray<BoundExpression> MakeStoreLoggerArguments(
+            ParameterSymbol parameter,
+            Symbol targetSymbol,
+            TypeSymbol targetType,
+            BoundExpression value,
+            BoundExpression? refAssignmentSourceIndex,
+            BoundExpression index)
+        {
+            if (refAssignmentSourceIndex != null)
+            {
+                return ImmutableArray.Create(_factory.Sequence(new[] { value }, refAssignmentSourceIndex), index);
+            }
+
+            Debug.Assert(parameter.RefKind == RefKind.None);
+
+            if (parameter.Type.IsVoidPointer() && !targetType.IsPointerOrFunctionPointer())
+            {
+                // address of assigned value to be passed to LogStore*Unmanaged:
+                Debug.Assert(!parameter.Type.IsManagedTypeNoUseSiteDiagnostics);
+
+                var addressOf = value is BoundLocal or BoundParameter ?
+                    (BoundExpression)new BoundAddressOfOperator(_factory.Syntax, value, isManaged: false, parameter.Type) :
+                    _factory.Sequence(new[] { value }, new BoundAddressOfOperator(_factory.Syntax, VariableRead(targetSymbol), isManaged: false, parameter.Type));
+
+                return ImmutableArray.Create(addressOf, _factory.Sizeof(targetType), index);
+            }
+
+            if (parameter.Type.SpecialType == SpecialType.System_String && targetType.SpecialType != SpecialType.System_String)
+            {
+                var toStringMethod = GetWellKnownMethodSymbol(WellKnownMember.System_Object__ToString, value.Syntax);
+
+                BoundExpression toString;
+                if (toStringMethod is null)
```

Comment:
>toStringMethod is null [](http://example.com/codeflow?start=20&length=22)

When will the well-known member be null?

Reply:
In practice never. In theory when building against bad corlib that doesn't have Object.ToString.